src/arcaflow_plugin_sdk/atp.py

```diff
@@ -50,73 +64,144 @@ class HelloMessage:
+            if work_start_msg["id"] is None:
+                self.stderr.write("Work start message is missing the 'id' field.")
+                return 1
+            if work_start_msg["config"] is None:
+                self.stderr.write("Work start message is missing the 'config' field.")
+                return 1
+
+            # Run the step
+            original_stdout = sys.stdout
+            original_stderr = sys.stderr
+            out_buffer = io.StringIO()
+            sys.stdout = out_buffer
+            sys.stderr = out_buffer
+            # Run the read loop
+            read_thread = threading.Thread(target=self.run_server_read_loop, args=(
+                plugin_schema,  # Plugin schema
+                work_start_msg["id"],  # step ID
+                decoder,  # Decoder
+            ))
+            read_thread.start()
+
+            output_id, output_data = plugin_schema.call_step(
+                work_start_msg["id"],
+                plugin_schema.unserialize_step_input(work_start_msg["id"], work_start_msg["config"])
+            )
+
+            # Send WorkDoneMessage in a RuntimeMessage
+            encoder.encode(
+                {
+                    "id": MessageType.WORK_DONE.value,
+                    "data": {
+                        "output_id": output_id,
+                        "output_data": plugin_schema.serialize_output(
+                            work_start_msg["id"], output_id, output_data
+                        ),
+                        "debug_logs": out_buffer.getvalue(),
+                    }
+                }
+            )
+            self.stdout.flush()  # Sends it to the ATP client immediately. Needed so it can realize it's done.
+            read_thread.join()  # Wait for the read thread to finish.
+            # Don't reset stdout/stderr until after the read thread is done.
+            sys.stdout = original_stdout
+            sys.stderr = original_stderr
+        except SystemExit:
             return 1
-        original_stdout = sys.stdout
-        original_stderr = sys.stderr
-        out_buffer = io.StringIO()
-        sys.stdout = out_buffer
-        sys.stderr = out_buffer
-        output_id, output_data = s.call_step(
-            message["id"], s.unserialize_input(message["id"], message["config"])
-        )
-        sys.stdout = original_stdout
-        sys.stderr = original_stderr
-        encoder.encode(
-            {
-                "output_id": output_id,
-                "output_data": s.serialize_output(
-                    message["id"], output_id, output_data
-                ),
-                "debug_logs": out_buffer.getvalue(),
-            }
-        )
-        stdout.flush()
-    except SystemExit:
-        return 1
-    return 0
+        return 0
+
+    def run_server_read_loop(
+        self,
+        plugin_schema: schema.SchemaType,
+        step_id: str,
+        decoder: cbor2.decoder.CBORDecoder,
+    ) -> None:
+        try:
+            while True:
```

Comment:
Concurrency does affect how `coverage` needs to observe the system under test. I have attempted to account for the program's concurrency and have not seen an increase `atp.py`'s coverage.

I added `.coveragerc` to the project root because `coverage` needs a configuration file to pass arguments to concurrent subroutines.
`
[run]
concurrency=multiprocessing,thread
`

Reply:
I added the coverage config file to the project.

Reply:
https://github.com/arcalot/arcaflow-plugin-sdk-python/issues/101 Issue created for coverage detection.